### PR TITLE
Add an Identifier by default

### PR DESCRIPTION
**Pull request by rynowak**
Using the MVC view engine convention for identifiers seems to make the
most sense, and we already use that convention for RazorProject so I
guess we're stuck with it.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs [hunk 1]
```diff
@@ -3,62 +3,12 @@
...
 
 namespace Microsoft.AspNetCore.Razor.Language
 {
     public static class RazorCodeDocumentExtensions
     {
-        // Internal for testing
-        internal static readonly string IdentifierKey = "identifier";
-        internal static readonly string ImportIdentifiersKey = "imports-identifiers";
-
-        public static string GetIdentifier(this RazorCodeDocument document)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs]: Now that we have relative paths on the source document, it just makes a lot more sense to compute the identifier on demand. I wasn't a big fan of putting it on the code document, and it's possible/easy to have a mismatch between the imports documents and identifiers.

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultMetadataIdentifierFeature.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                throw new ArgumentNullException(nameof(sourceDocument));
+            }
+
+            if (string.IsNullOrEmpty(sourceDocument.RelativePath))
+            {
+                return null;
+            }
+
+            var identifier = sourceDocument.RelativePath;
+            if (!identifier.StartsWith("/") && !identifier.StartsWith("\\"))
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultMetadataIdentifierFeature.cs]: Could get rid of the second part of this `if` if you `Replace("\\", "/")` first.

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    internal interface IMetadataIdentifierFeature : IRazorEngineFeature
+    {
+        string GetIdentifier(RazorCodeDocument codeDocument, RazorSourceDocument sourceDocument);
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs]: Is `RazorCodeDocument` even necessary in this api?

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultMetadataIdentifierFeature.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultMetadataIdentifierFeature.cs]: Yeah that might be clearer 👍

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs (hunk 3 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs]: It's not necessary in the default implementation

Comment (ajaybhargavb) [reply src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs]: Oops I meant `RazorSourceDocument`. Why not use `codeDocument.Source`?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs]: This applies to imports too

Comment (ajaybhargavb) [reply src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataIdentifierFeature.cs]: Ah. Got it 👍

**Comment by rynowak** [comment]
Looking into the appveyor failure. That's not related to this change at all

**Pull request merged**